### PR TITLE
initial collator selection

### PR DESCRIPTION
**Pull request by GopherJ**
close: #117 

This uses the same method as statemint, acala has different method

1. anyone can register himself as candidate using `register_as_candidate`, some tokens will be reserved
2. fees will be collected by collator selection's pallet account
3. every session will kick stale candidates, stale candidates => not authoring block in N blocks
4. author of block can win 1/2 of collator selection's pallet account balance - min_balance (ED)

now collator can get native token, in the future they should be able to get PARA/HEIKO

- [x] test producing block still works
- [x] test register as collator and win fees

### CODE REVIEW

File: node/parallel/src/chain_spec/heiko.rs [hunk 1]
```diff
@@ -161,10 +164,25 @@ fn testnet_genesis(
                     .collect()
             },
         },
-        // TODO collateral selection
-        pallet_aura: AuraConfig {
-            authorities: initial_authorities.iter().map(|x| (x.1.clone())).collect(),
+        pallet_collator_selection: CollatorSelectionConfig {
+            invulnerables: invulnerables.iter().cloned().map(|(acc, _)| acc).collect(),
+            candidacy_bond: EXISTENTIAL_DEPOSIT * 16,
```
Comment (kaichaosun) [inline node/parallel/src/chain_spec/heiko.rs:169]: There is a potential bug in this pallet, `candidacy_bond` is not reserved in genesis config.

File: runtime/heiko/src/lib.rs [hunk 2]
```diff
@@ -374,14 +376,65 @@ impl pallet_timestamp::Config for Runtime {
...
+parameter_types! {
+    pub const PotId: PalletId = PalletId(*b"PotStake");
+    pub const MaxCandidates: u32 = 1000;
+    pub const MaxInvulnerables: u32 = 100;
+}
+
+impl pallet_collator_selection::Config for Runtime {
+    type Event = Event;
+    type Currency = Balances;
+    type UpdateOrigin = EnsureRoot<AccountId>;
```
Comment (kaichaosun) [inline runtime/heiko/src/lib.rs:421]: Half of council should also be fine.

**Comment by GopherJ** [comment]
it worked finally:

image

File: node/parallel/src/chain_spec/heiko.rs (hunk 1 above)
Comment (GopherJ) [reply node/parallel/src/chain_spec/heiko.rs:169]: @kaichaosun can you explain more on this? normally `candidacy_bond` will only be used while registering a new candidate

**Comment by GopherJ** [comment]
@kaichaosun 

Maxcandidates is used only for measuring weight:

https://github.com/paritytech/cumulus/blob/master/pallets/collator-selection/src/lib.rs#L294

File: runtime/heiko/src/constants.rs [hunk 3]
```diff
@@ -14,6 +14,9 @@
 
 pub mod currency {
     use primitives::Balance;
+
+    pub const EXISTENTIAL_DEPOSIT: u128 = 500;
+
```
Comment (ayushmishra2005) [inline runtime/heiko/src/constants.rs:19]: Can we remove empty line?

File: runtime/parallel/src/constants.rs [hunk 4]
```diff
@@ -14,6 +14,9 @@
 
 pub mod currency {
     use primitives::Balance;
+
+    pub const EXISTENTIAL_DEPOSIT: u128 = 500;
+
```
Comment (ayushmishra2005) [inline runtime/parallel/src/constants.rs:19]: Can we remove empty line?

File: node/parallel/src/chain_spec/heiko.rs (hunk 1 above)
Comment (kaichaosun) [reply node/parallel/src/chain_spec/heiko.rs:169]: Are there more context about this one?

**Review by kaichaosun — COMMENTED** [review]
There are still unresolved questions, let's slow the merge process a bit.

File: runtime/heiko/src/constants.rs (hunk 3 above)
Comment (GopherJ) [reply runtime/heiko/src/constants.rs:19]: > There are still unresolved questions, let's slow the merge process a bit.

I think this one doesn't have any security issue, it's just how shall we use it.

**Comment by GopherJ** [comment]
> There are still unresolved questions, let's slow the merge process a bit.

I think this one doesn't have any security issue, it's just how should we use it, the initial invulnerables should be our collators

File: pallets/loans/src/tests/liquidate_borrow.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
+use crate::{
```
Comment (kaichaosun) [inline pallets/loans/src/tests/liquidate_borrow.rs:1]: Is this from other commits?

File: pallets/loans/src/tests.rs [hunk 6]
```diff
@@ -79,13 +81,20 @@ fn mint_works() {
 }
 
 #[test]
-fn mint_failed() {
+fn mint_must_return_err_when_overflows_occur() {
```
Comment (kaichaosun) [inline pallets/loans/src/tests.rs:84]: And this not related to this PR.

Comment (GopherJ) [reply pallets/loans/src/tests.rs:84]: I think this is because that I rebased the master

File: pallets/loans/src/tests/liquidate_borrow.rs (hunk 5 above)
Comment (GopherJ) [reply pallets/loans/src/tests/liquidate_borrow.rs:1]: yes it's probably be caused by rebase

**Review by kaichaosun — APPROVED** [review]
Well done. 👍

**Pull request merged**